File: GLSS_convert_sav.py
import os
import pandas as pd

folder_paths = [
    "2005 with csvs/aggregates", "2005 with csvs/parta", "2005 with csvs/section10",
    "1991-1992/aggreg", "1991-1992/Prices", "1998/aggreg", "1998/Prices",
    "2012-2013/AGGREGATES", "2012-2013/PARTA", "2012-2013/PRICES",
    "2017/g7aggregates", "2017/g7PartA", "2017/g7price"
]

for folder in folder_paths:
    if not os.path.exists(folder):
        print(f"Skipping: Folder does not exist - {folder}")
        continue

    csv_folder = f"{folder}_csv"
    os.makedirs(csv_folder, exist_ok=True)

    all_files = [f for f in os.listdir(folder) if os.path.isfile(os.path.join(folder, f))]
    valid_extensions = ["sav", "dta", "SAV", "DTA"]

    data_files = [f for f in all_files if f.split('.')[-1].lower() in valid_extensions]

    if not data_files:
        print(f"No .sav or .dta files found in {folder}")
        continue

    for file in data_files:
        file_path = os.path.join(folder, file)
        file_ext = file.split('.')[-1].lower()
        data = None

        print(f"Processing: {file_path}")

        try:
            if file_ext == "sav":
                print(f"Trying pandas read_spss for {file_path}")
                data = pd.read_spss(file_path)
            elif file_ext == "dta":
                print(f"Trying pandas read_stata for {file_path}")
                data = pd.read_stata(file_path)

            if data is not None and not data.empty:
                csv_file = os.path.join(csv_folder, f"{os.path.splitext(file)[0]}.csv")
                data.to_csv(csv_file, index=False)
                print(f"Successfully converted: {file_path} -> {csv_file}")
            else:
                print(f"Skipping: No data found in {file_path}")
        except Exception as e:
            print(f"Failed to process {file_path}: {e}")

print("All files processed.")

File: GLSS_convert_sav.r
library(haven)
library(foreign)
library(rio)
library(readr)

folder_paths <- c("2005 with csvs/aggregates","2005 with csvs/parta", "2005 with csvs/section10", "1991-1992/aggreg",  
                  "1991-1992/Prices", "1998/aggreg", "1998/Prices","2012-2013/AGGREGATES", "2012-2013/PARTA", "2012-2013/PRICES", 
                  "2017/g7aggregates", "2017/g7PartA", "2017/g7price")

for (folder in folder_paths) {
  if (!dir.exists(folder)) {
    message(paste("Skipping: Folder does not exist -", folder))
    flush.console()
    next
  }

  csv_folder <- paste0(folder, "_csv")
  dir.create(csv_folder, recursive = TRUE, showWarnings = FALSE)

  all_files <- list.files(folder, full.names = TRUE)
  valid_extensions <- c("sav", "dta")
  data_files <- all_files[tolower(tools::file_ext(all_files)) %in% valid_extensions]

  if (length(data_files) == 0) {
    message(paste("No .sav or .dta files found in", folder))
    flush.console()
    next
  }

  for (file in data_files) {
    data <- NULL  

    message(paste("Processing:", file))
    flush.console()

    file_ext <- tolower(tools::file_ext(file))

    if (file_ext == "sav") {
      tryCatch({
        message(paste("Trying haven for", file))
        flush.console()
        data <- read_sav(file, user_na = TRUE)
      }, error = function(e) {
        message(paste("Haven failed for", file, "-", e$message))
        flush.console()
      })

      if (is.null(data) || nrow(data) == 0) {
        tryCatch({
          message(paste("Trying foreign for", file))
          flush.console()
          data <- read.spss(file, to.data.frame = TRUE, use.value.labels = FALSE)
        }, error = function(e) {
          message(paste("Foreign failed for", file, "-", e$message))
          flush.console()
        })
      }
    }

    if (file_ext == "dta") {
      tryCatch({
        message(paste("Trying haven for", file))
        flush.console()
        data <- read_dta(file)
      }, error = function(e) {
        message(paste("Haven failed for", file, "-", e$message))
        flush.console()
      })
    }

    if (is.null(data) || nrow(data) == 0) {
      tryCatch({
        message(paste("Trying rio for", file))
        flush.console()
        data <- import(file)
      }, error = function(e) {
        message(paste("Rio failed for", file, "-", e$message))
        flush.console()
      })
    }

    if (is.null(data) || nrow(data) == 0) {
      message(paste("Skipping: No data found in", file))
      flush.console()
      next
    }

    print(head(data))
    flush.console()

    csv_file <- file.path(csv_folder, paste0(tools::file_path_sans_ext(basename(file)), ".csv"))
    write_csv(data, csv_file)

    message(paste("Successfully converted:", file, "->", csv_file))
    flush.console()
  }
}

message("All files processed.")
flush.console()
